src/main/scala/firrtl/passes/wiring/Wiring.scala

```diff
@@ -0,0 +1,165 @@
+    * Notes:
+    *   - No module uniquification occurs (due to imposed restrictions)
+    */
+  def run(c: Circuit): Circuit = {
+    // Split out WiringInfo
+    val source = wi.source
+    val sinks = wi.sinks.keys.toSet
+    val compName = wi.comp
+
+    // Maps modules to children instances, i.e. (instance, module)
+    val childrenMap = getChildrenMap(c)
+
+    // Check restrictions
+    val nSources = countInstances(childrenMap, wi.top, source)
+    if(nSources != 1)
+      throw new WiringException(s"Cannot have $nSources instance of $source under ${wi.top}")
+    sinks.foreach { m =>
+      val total = countInstances(childrenMap, c.main, m)
+      val nTop = countInstances(childrenMap, c.main, wi.top)
+      val perTop = countInstances(childrenMap, wi.top, m)
+      if(total != nTop * perTop)
+        throw new WiringException(s"Module ${wi.top} does not contain all instances of $m.")
+    }
+
+    // Create valid port names for wiring that have no name conflicts
+    val portNames = c.modules.foldLeft(Map[String, String]()){(map, m) =>
+      map + (m.name -> {
+        val ns = Namespace(m)
+        wi.sinks.get(m.name) match {
+          case Some(pin) => ns.newName(pin)
+          case None => ns.newName(tokenize(compName).filter(t => !("[].".contains(t))).reduce(_ + "_" + _))
+        }
+      })
+    }
+
+    // Create a lineage tree from children map
+    val lineages = getLineage(childrenMap, wi.top)
+
+    // Populate lineage tree with relationship information, i.e. who is source,
+    //   sink, parent of source, etc.
+    val withFields = setSharedParent(wi.top)(setFields(sinks, source)(lineages))
+
+    // Populate lineage tree with what to instantiate, connect to/from, etc.
+    val withThings = setThings(portNames, compName)(withFields)
+
+    // Create a map from module name to lineage information
+    val map = pointToLineage(withThings)
+
+    // Obtain the source component type
+    val sourceComponentType = getType(c, source, compName)
+
+    // Return new circuit with correct wiring
+    val cx = c.copy(modules = c.modules map onModule(map, sourceComponentType))
+
+    // Replace inserted IR nodes with WIR nodes
+    ToWorkingIR.run(cx)
+  }
+
+  /** Return new module with correct wiring
+    */
+  private def onModule(map: Map[String, Lineage], t: Type)(m: DefModule) = {
+    map.get(m.name) match {
+      case None => m
+      case Some(l) =>
+        val stmts = mutable.ArrayBuffer[Statement]()
+        val ports = mutable.ArrayBuffer[Port]()
+        l.addPort match {
+          case None =>
+          case Some((s, dt)) => dt match {
+            case DecInput => ports += Port(NoInfo, s, Input, t)
+            case DecOutput => ports += Port(NoInfo, s, Output, t)
+            case DecWire =>
+              stmts += DefWire(NoInfo, s, t)
+          }
+        }
+        stmts ++= l.cons.map{case ((l, r)) => 
+          Connect(NoInfo, toExp(l), toExp(r))
+        }
```

Comment:
space around { and I would probably use a space instead of a dot on the map here